### PR TITLE
Improve diffing of collections and structured values.

### PR DESCRIPTION
**Pull request by grynspan**
This PR improves the test output when comparing collections and structured values (using `#expect()`) when the comparison fails. Previously, we'd just the insertions and deletions as arrays, which was potentially hard to read and didn't show the test author _where_ these changes occurred. Now, we convert the compared values into something similar to multi-line `diff` output. Consider the following test function:

```swift
@Test func f() {
  let lhs = """
  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
  tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
  quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
  consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
  cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
  non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  """
  let rhs = """
  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
  tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
  quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
  potato salad. Duis aute irure dolor in reprehenderit in voluptate velit esse
  cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
  non dentistry, sunt in culpa qui officia deserunt mollit anim id est laborum.
  """
  #expect(lhs == rhs)
}
```

The output will now be:

```
◇ Test f() started.
✘ Test f() recorded an issue at MyTests.swift:45:5: Expectation failed: lhs == rhs
± 2 changes:
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod"
  "tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,"
  "quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo"
+ "potato salad. Duis aute irure dolor in reprehenderit in voluptate velit esse"
- "consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse"
  "cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat"
+ "non dentistry, sunt in culpa qui officia deserunt mollit anim id est laborum."
- "non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
✘ Test f() failed after 0.001 seconds with 1 issue.
```

Similarly, with this test:

```swift
struct S: Equatable {
  var x: Int
  var y: String
}

@Test func f() {
  let lhs = S(x: 123, y: "abc")
  let rhs = S(x: 123, y: "def")
  #expect(lhs == rhs)
}
```

The output should appear similar to:

```
◇ Test f() started.
✘ Test f() recorded an issue at MyTests.swift:36:5: Expectation failed: lhs == rhs
± 1 change:
  MyTests.S(
    x: 123
-   y: "abc"
+   y: "def"
  )
✘ Test f() failed after 0.001 seconds with 1 issue.
```

Resolves rdar://66351980.

### CODE REVIEW

**Comment by SeanROlszewski** [comment]
So many thank yous are in order for you, @grynspan.

**Comment by SeanROlszewski** [comment]
It's probably worth updating the docs for this change to mention that the diffing logic uses `CustomStringConvertible` and `CustomDebugStringConvertible` via `String(describing:)` under the hood, which can lead to unexpected diffs if the user does something unique or hard to diff.

**Comment by grynspan** [comment]
Still working on the exact semantics of this change, since the exploded view of a value is not always the _right_ view for comparison. But yes, we'll want to amend documentation at the same time.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

File: Sources/Testing/Events/Event.Recorder.swift [hunk 1]
```diff
@@ -344,6 +344,37 @@ extension Event.Recorder {
...
+  fileprivate func formattedDescription(options: Set<Event.Recorder.Option>) -> String {
+    guard options.contains(.useANSIEscapeCodes) else {
+      return String(describing: self)
+    }
+
+    return String(describing: self)
+      .split(whereSeparator: \.isNewline)
+      .map { line in
+        switch line.first {
+        case "+":
```
Comment (grynspan) [inline Sources/Testing/Events/Event.Recorder.swift:365]: I'm thinking about refactoring this code a bit in the future to use SF Symbols when available, but I want to factor `_Symbol` out into its own file first, and that's not relevant to this PR.

**Comment by grynspan** [comment]
@swift-ci please test

**Pull request closed without merging**